### PR TITLE
Optional Trip Dispatcher params

### PR DESCRIPTION
**Pull request by tildeee**
It'll start to be painful to unit test Wave 3. The student needs to unit test the following things:

- Have a set of Drivers who are available and unavailable,
- of the available drivers, have some who have more than one trip associated with them
  - of these trips, know which trip is most recent
    - of those trips, know which trip is oldest, and pick the driver associated with them

Unless we find that driver in the driver CSV, it'll be pretty goofy for them to work with those tests.

This is a proposal for optional params. With optional params, in tests the students could pass in their own arrays of test drivers and test passengers, which makes the test an isolated way to give their own data.

The thing to decide is:

with these changes: we shouldn't load trips from the CSV if drivers or passengers was passed in, and make the trips array empty in that case.

However, the requirements for assigning drivers tells the students to only consider the trips for each driver who have any trips at all (it assumes each driver has at least one trip).

Of course, in a test, you can just make a test trip and give it to the test drivers, etc, but that requires making Trip Dispatcher having the other parts of "creating a trip" set correctly... which we aren't providing.

for example, the ideal test would be:
```ruby
it "should create a trip that picks the right driver" do
  right_driver = RideShare::Driver.new( ... )
  wrong_driver = RideShare::Driver.new( ... )
  passenger = RideShare::Passenger.new( ... )

  dispatcher = RideShare::Dispatcher.new( [ right_driver, wrong_driver ], [ passenger ] )

  new_trip = dispatcher.create_trip_for_passenger( passenger.id )

  new_trip.driver.must_be right_driver
end
```

But as is, we'd have to do this:
```ruby
it "should create a trip that picks the right driver" do
  right_driver = RideShare::Driver.new( ... )
  wrong_driver = RideShare::Driver.new( ... )
  trip_a = RideShare::Trip.new( ... )
  trip_b = RideShare::Trip.new( ... )
  trip_c = RideShare::Trip.new( ... )
  trip_d = RideShare::Trip.new( ... )
  right_driver.add_trip(trip_a)
  right_driver.add_trip(trip_b)
  wrong_driver.add_trip(trip_c)
  wrong_driver.add_trip(trip_d)

  passenger = RideShare::Passenger.new( ... )
  # optional but...
  passenger.add_trip(trip_a)
  passenger.add_trip(trip_b)
  passenger.add_trip(trip_c)
  passenger.add_trip(trip_d)
  

  dispatcher = RideShare::Dispatcher.new( [ right_driver, wrong_driver ], [ passenger ] )
  dispatcher.trips << trip_a # ... etc

  new_trip = dispatcher.create_trip_for_passenger( passenger.id )

  new_trip.driver.must_be right_driver
end
```

That's where I'm worried and stuck! OK THANKS

### CODE REVIEW

**Comment by kariabancroft** [comment]
If the logic has to look at the trips list to ensure that the logic defined works, then how would the optional params with no trips work? Wouldn't it pick no driver since they don't have trips? Or are you also proposed we remove/adjust this requirement?

File: lib/trip_dispatcher.rb [hunk 1]
```diff
@@ -9,10 +9,10 @@ module RideShare
   class TripDispatcher
     attr_reader :drivers, :passengers, :trips
 
-    def initialize
-      @drivers = load_drivers
-      @passengers = load_passengers
-      @trips = load_trips
+    def initialize(drivers=nil, passengers=nil)
+      @drivers = drivers != nil ? drivers : load_drivers
```
Comment (droberts-ada) [inline lib/trip_dispatcher.rb]: We should use keyword args for these parameters!

**Review by droberts-sea — COMMENTED** [review]
I definitely agree that testing wave 3 is going to be a bear, and I think this patch makes it a much smaller bear. But, it's still a bear.

I'm not sure there's much more we _can_ do - that's part of the challenge of working with complex structures. Maybe add (or recommend they add) a method like this:

```ruby
class TripDispatcher
  # Adds a completed trip and sets up all relations
  def add_trip(id, driver_id, passenger_id, start_time: (Time.now-3600), end_time: (Time.now-1800), cost: 17.25, rating: 3)
    # basically the inner loop of find_trip
    driver = find_driver(driver_id)
    passenger = find_passenger(passenger_id)
    trip = Trip.new( ... )
    @trips << trip
    driver.add_trip(trip)
    passenger.add_trip(trip)
  end
end
```

That way they don't have to set up the relations themselves
